File: backend/Services/AlunoService.cs
using Microsoft.EntityFrameworkCore;
using NeuroGest.API.Data;
using NeuroGest.API.DTOs;
using NeuroGest.API.Models;

namespace NeuroGest.API.Services;

public class AlunoService : IAlunoService
{
    private readonly AppDbContext _db;

    public AlunoService(AppDbContext db)
    {
        _db = db;
    }

    // ─── Listar (com busca opcional) ──────────────────────
    public async Task<List<AlunoResponseDto>> ListarAsync(string? busca = null)
    {
        var query = _db.Alunos.AsQueryable();

        if (!string.IsNullOrWhiteSpace(busca))
        {
            busca = busca.ToLower();
            query = query.Where(a =>
                a.Nome.ToLower().Contains(busca) ||
                (a.NomePai != null && a.NomePai.ToLower().Contains(busca)) ||
                (a.NomeMae != null && a.NomeMae.ToLower().Contains(busca)));
        }

        var lista = await query.OrderBy(a => a.Nome).ToListAsync();
        return lista.Select(ToDto).ToList();
    }

    // ─── Buscar por ID ────────────────────────────────────
    public async Task<AlunoResponseDto?> BuscarPorIdAsync(int id)
    {
        var aluno = await _db.Alunos.FindAsync(id);
        return aluno is null ? null : ToDto(aluno);
    }

    // ─── Criar ────────────────────────────────────────────
    public async Task<AlunoResponseDto> CriarAsync(AlunoDto dto)
    {
        var aluno = new Aluno
        {
            Nome                = dto.Nome.Trim(),
            DataNascimento      = dto.DataNascimento,
            Municipio           = dto.Municipio?.Trim(),
            Sexo                = dto.Sexo?.Trim(),
            Estado              = dto.Estado?.Trim(),
            NomePai             = dto.NomePai?.Trim(),
            NomeMae             = dto.NomeMae.Trim(),
            Cpf                 = dto.Cpf.Trim(),
            CpfPai              = dto.CpfPai?.Trim(),
            CpfMae              = dto.CpfMae?.Trim(),
            TelefoneResponsavel = dto.TelefoneResponsavel?.Trim(),
            Observacoes         = dto.Observacoes?.Trim(),
        };

        _db.Alunos.Add(aluno);
        await _db.SaveChangesAsync();
        return ToDto(aluno);
    }

    // ─── Editar ───────────────────────────────────────────
    public async Task<(AlunoResponseDto? dto, string? erro)> EditarAsync(int id, AlunoDto dto)
    {
        var aluno = await _db.Alunos.FindAsync(id);
        if (aluno is null)
            return (null, "Aluno não encontrado.");

        aluno.Nome                = dto.Nome.Trim();
        aluno.DataNascimento      = dto.DataNascimento;
        aluno.Municipio           = dto.Municipio?.Trim();
        aluno.Sexo                = dto.Sexo?.Trim();
        aluno.Estado              = dto.Estado?.Trim();
        aluno.NomePai             = dto.NomePai?.Trim();
        aluno.NomeMae             = dto.NomeMae.Trim();
        aluno.Cpf                 = dto.Cpf.Trim();
        aluno.CpfPai              = dto.CpfPai?.Trim();
        aluno.CpfMae              = dto.CpfMae?.Trim();
        aluno.TelefoneResponsavel = dto.TelefoneResponsavel?.Trim();
        aluno.Observacoes         = dto.Observacoes?.Trim();

        await _db.SaveChangesAsync();
        return (ToDto(aluno), null);
    }

    // ─── Ativar / Desativar ───────────────────────────────
    public async Task<bool> AlternarAtivoAsync(int id)
    {
        var aluno = await _db.Alunos.FindAsync(id);
        if (aluno is null) return false;

        aluno.Ativo = !aluno.Ativo;
        await _db.SaveChangesAsync();
        return true;
    }

    // ─── Excluir ──────────────────────────────────────────
    public async Task<(bool ok, string? erro)> ExcluirAsync(int id)
    {
        var aluno = await _db.Alunos.FindAsync(id);
        if (aluno is null) return (false, "Aluno não encontrado.");

        _db.Alunos.Remove(aluno);
        try
        {
            await _db.SaveChangesAsync();
        }
        catch (DbUpdateException)
        {
            return (false, "Não é possível excluir: existem atendimentos, agendamentos ou lançamentos vinculados a este aluno.");
        }
        return (true, null);
    }

    // ─── Mapeamento ───────────────────────────────────────
    private static AlunoResponseDto ToDto(Aluno a) => new()
    {
        Id                  = a.Id,
        Nome                = a.Nome,
        DataNascimento      = a.DataNascimento,
        Idade               = a.Idade,
        Municipio           = a.Municipio,
        Sexo                = a.Sexo,
        Estado              = a.Estado,
        NomePai             = a.NomePai,
        NomeMae             = a.NomeMae,
        Cpf                 = a.Cpf,
        CpfPai              = a.CpfPai,
        CpfMae              = a.CpfMae,
        TelefoneResponsavel = a.TelefoneResponsavel,
        Observacoes         = a.Observacoes,
        Ativo               = a.Ativo,
        CriadoEm           = a.CriadoEm,
    };
}

File: frontend/lib/services/aluno_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/aluno.dart';
import 'auth_service.dart';

class AlunoService {
  static const String _baseUrl = 'https://api.neurogest.online';

  static Future<Map<String, String>> _headers() async {
    final token = await AuthService.getToken();
    return {
      'Content-Type': 'application/json',
      'Authorization': 'Bearer $token',
    };
  }

  // ─── Listar ───────────────────────────────────────────
  static Future<List<Aluno>> listar({String? busca}) async {
    final uri = Uri.parse('$_baseUrl/api/alunos')
        .replace(queryParameters: busca != null && busca.isNotEmpty ? {'busca': busca} : null);

    final response = await http.get(uri, headers: await _headers());

    if (response.statusCode == 200) {
      final List<dynamic> lista = jsonDecode(response.body);
      return lista.map((j) => Aluno.fromJson(j)).toList();
    }
    return [];
  }

  // ─── Buscar por ID ────────────────────────────────────
  static Future<Aluno?> buscarPorId(int id) async {
    final response = await http.get(
      Uri.parse('$_baseUrl/api/alunos/$id'),
      headers: await _headers(),
    );
    if (response.statusCode == 200) {
      return Aluno.fromJson(jsonDecode(response.body));
    }
    return null;
  }

  // ─── Criar ────────────────────────────────────────────
  static Future<String?> criar({
    required String nome,
    required String dataNascimento,
    String? municipio,
    String? sexo,
    String? estado,
    String? nomePai,
    required String nomeMae,
    required String cpf,
    String? telefoneResponsavel,
    String? observacoes,
  }) async {
    final response = await http.post(
      Uri.parse('$_baseUrl/api/alunos'),
      headers: await _headers(),
      body: jsonEncode({
        'nome':                nome,
        'dataNascimento':      dataNascimento, // formato: yyyy-MM-dd
        'municipio':           municipio,
        'sexo':                sexo,
        'estado':              estado,
        'nomePai':             nomePai,
        'nomeMae':             nomeMae,
        'cpf':                 cpf,
        'telefoneResponsavel': telefoneResponsavel,
        'observacoes':         observacoes,
      }),
    );

    if (response.statusCode == 201) return null;
    try {
      final body = jsonDecode(response.body);
      return body['mensagem'] ?? 'Erro ao criar aluno.';
    } catch (_) {
      return 'Erro ao criar aluno.';
    }
  }

  // ─── Editar ───────────────────────────────────────────
  static Future<String?> editar({
    required int id,
    required String nome,
    required String dataNascimento,
    String? municipio,
    String? sexo,
    String? estado,
    String? nomePai,
    required String nomeMae,
    required String cpf,
    String? telefoneResponsavel,
    String? observacoes,
  }) async {
    final response = await http.put(
      Uri.parse('$_baseUrl/api/alunos/$id'),
      headers: await _headers(),
      body: jsonEncode({
        'nome':                nome,
        'dataNascimento':      dataNascimento,
        'municipio':           municipio,
        'sexo':                sexo,
        'estado':              estado,
        'nomePai':             nomePai,
        'nomeMae':             nomeMae,
        'cpf':                 cpf,
        'telefoneResponsavel': telefoneResponsavel,
        'observacoes':         observacoes,
      }),
    );

    if (response.statusCode == 200) return null;
    try {
      final body = jsonDecode(response.body);
      return body['mensagem'] ?? 'Erro ao editar aluno.';
    } catch (_) {
      return 'Erro ao editar aluno.';
    }
  }

  // ─── Excluir ──────────────────────────────────────────
  static Future<bool> excluir(int id) async {
    final response = await http.delete(
      Uri.parse('$_baseUrl/api/alunos/$id'),
      headers: await _headers(),
    );
    return response.statusCode == 200;
  }
}